### ISSUE TITLE
Better validation checks when adding a private registry

### ISSUE THREAD

**Issue by will-chan** [rancher/rancher#300]
Currently, if you add a private registry, there is no check that the registry is even valid. We should at least attempt to validate it, otherwise the UX is poor when attempting to use the image from that registry.
1. We should try to validate the registry URL is even valid or that we can connect to it?  Can't we do a simple listImage API command (or equivalent) to validate we can list images from the registry.
2. Can we also attempt a login into the registry to validate the credentials as well?

**Comment by vincent99** [comment]
This would have to be done on the API side, the UI can't make arbitrary requests (and may not be able to reach the private registry even if it would be accessible by the hosts).

We'd probably want to add something like a "testRegistry" collectionAction that doesn't require having created actual registry & credential resources, because that is 2 posts + wait for state to become active + cleanup if it fails.

**Comment by mikehancock** [comment]
+1 
In addition could you specify that only domain is required in Registry Address, that tripped me up first time round as I simply pasted the url

**Comment by vincent99** [comment]
I think we need a round of overhaul on how private registries work in general to make it usable.  @ibuildthecloud was thinking maybe we just have credentials you can choose from and send to whatever registry is in the image `docker:registry/namespace/name:tag` value.

**Comment by vincent99** [comment]
Update: we still need support for credential validation in the API to make this really friendly, but there are now convenient buttons for DockerHub and Quay.io that don't require you to know the registry address in rancherio/ui#152.

**Comment by hibooboo2** [comment]
I believe there are two options we can go for here:

1. Just strip the url provided to either the domain:port at the api or ui level. 
2. Before the registry is created in db cattle sends an event to a host that will have the host try to ping the domain and then send the response to cattle and respond appropriately.

1 is simple and quick, 2 would require adding ability to make arbitrary requests through python agent as well as mean you are only able to use a registry/create one for a url that some host can hit, but you don't know which host to test from. It may be a registry only one of the hosts can access.

**Comment by oxr463** [comment]
@vincent99 why was this closed? It doesn't seem like this was ever fixed.

**Comment by vincent99** [comment]
It's 5 1/2 years old, for (pre-)1.x, and things we were never going to do for it were closed around the release of 2.x.

TL;DR: We could as basic validation that it looks like a syntactically valid URL in the UI-only, but that's about it.

Kubernetes has pretty poorly-defined openapi validation rules on the schemas in general for guiding preemptive automatic checks (before submitting the form) in the UI to begin with.

And a "registry" is really just a secret with a particular set of key/values, so any arbitrary data is a perfectly valid secret as far as the k8s API cares.

Enforcing this properly would require intercepting the request from the k8s api on all downstream clusters, probably doing blocking network requests to see if we like that name or not (which now becomes a DOS attack vector), and then deciding to pass it through or not. And still has the same problems as above like maybe it's valid from a worker node but not from the control plane. So this is just not going to happen.
